### PR TITLE
Use Char instead of String at escape sequences

### PR DESCRIPTION
**Pull request by r00ster91**
And some other improvements like using [String#delete](https://crystal-lang.org/api/0.24.2/String.html#delete%28char%3AChar%29-instance-method) instead of [String#gsub](https://crystal-lang.org/api/0.24.2/String.html#gsub%28char%3AChar%2Creplacement%29-instance-method) when it's something like `gsub(' ', "")`.

### CODE REVIEW

File: src/compiler/crystal/tools/doc/macro.cr [hunk 1]
```diff
@@ -39,7 +39,7 @@ class Crystal::Doc::Macro
   end
 
   def anchor
-    "#" + URI.escape(id)
+    '#' + URI.escape(id)
```
Comment (ysbaddaden) [inline src/compiler/crystal/tools/doc/macro.cr]: This one I'm not sure sure : Char + String doesn't look quite good.

File: src/compiler/crystal/compiler.cr [hunk 2]
```diff
@@ -441,9 +441,7 @@ module Crystal
       puts
       puts "Macro runs:"
       program.compiled_macros_cache.each do |filename, compiled_macro_run|
-        print " - "
-        print filename
-        print ": "
+        print " - #{filename}: "
```
Comment (ysbaddaden) [inline src/compiler/crystal/compiler.cr]: Not very important in this case (not a potential tight loop) but the original 3 prints pushed 2 static strings and 1 dynamically allocated string to the output buffer, whereas your change now allocates another dynamic string and eventually pushes it to the output buffer, hence requires a thrown-away allocation.

File: src/compiler/crystal/tools/doc/macro.cr (hunk 1 above)
Comment (jhass) [reply src/compiler/crystal/tools/doc/macro.cr]: Well, we should make it good if it's bad I guess.

File: src/compiler/crystal/compiler.cr (hunk 2 above)
Comment (r00ster91) [reply src/compiler/crystal/compiler.cr]: I also made a benchmark
```Crystal
require "benchmark"

filename = "file.cr"

Benchmark.ips do |x|
  x.report "3 prints" do
    print " - "
    print filename
    print ": "
  end

  x.report "1 print" do
    print " - #{filename}: "
  end
end
```
and the 1 print is 2 times faster.
```
3 prints   8.35k (119.77µs) (± 6.16%)  2.30× slower
 1 print  19.17k ( 52.16µs) (± 2.83%)       fastest
```

File: spec/compiler/crystal/tools/implementations_spec.cr [hunk 3]
```diff
@@ -25,7 +25,7 @@ private def assert_implementations(code)
     end
   end
 
-  code = code.gsub('‸', "").gsub('༓', "")
+  code = code.delete('‸').delete('༓')
```
Comment (j8r) [inline spec/compiler/crystal/tools/implementations_spec.cr]: `code.delete { |c|  {'‸', '༓'}.includes? c }` may yield to more performance.

File: src/compiler/crystal/tools/doc/macro.cr [hunk 4]
```diff
@@ -29,7 +29,7 @@ class Crystal::Doc::Macro
 
   def id
     String.build do |io|
-      io << to_s.gsub(/<.+?>/, "").gsub(' ', "")
+      io << to_s.gsub(/<.+?>/, "").delete(' ')
       io << "-macro"
```
Comment (j8r) [inline src/compiler/crystal/tools/doc/macro.cr]: `io << to_s.gsub(/<.+?>/, "").delete(' ') << "-macro"` on a single line

Comment (j8r) [inline src/compiler/crystal/tools/doc/macro.cr]: `io << to_s.gsub(/<.+?>/, "").delete(' ') << "-macro"` on a single line

File: spec/compiler/crystal/tools/implementations_spec.cr (hunk 3 above)
Comment (r00ster91) [reply spec/compiler/crystal/tools/implementations_spec.cr]: Yep it's 2 times faster.

File: src/compiler/crystal/compiler.cr (hunk 2 above)
Comment (bew) [reply src/compiler/crystal/compiler.cr]: What the.... Why is that faster? @ysbaddaden you know?

Comment (hugoabonizio) [reply src/compiler/crystal/compiler.cr]: It's related to IO being flushed, if you use another function call IO stops being the bottleneck:

```crystal
require "benchmark"

def print2(param); end

filename = "file.cr"

Benchmark.ips do |x|
  x.report "3 prints" do
    print2 " - "
    print2 filename
    print2 ": "
  end

  x.report "1 print" do
    print2 " - #{filename}: "
  end
end
```

```
3 prints 442.72M (  2.26ns) (± 4.78%)       fastest
 1 print   6.52M (153.48ns) (± 4.70%) 67.95× slower
```

Comment (straight-shoota) [reply src/compiler/crystal/compiler.cr]: Fastest is probably `print " - ", filename, ": "`

File: src/compiler/crystal/tools/doc/macro.cr [hunk 5]
```diff
@@ -29,8 +29,7 @@ class Crystal::Doc::Macro
 
   def id
     String.build do |io|
-      io << to_s.gsub(/<.+?>/, "").gsub(' ', "")
-      io << "-macro"
+      io << to_s.gsub(/<.+?>/, "").delete(' ') << "-macro"
```
Comment (straight-shoota) [inline src/compiler/crystal/tools/doc/macro.cr]: Pulling `<< "-macro"` into the previous line is an unrelated change. It doesn't really improve performance like the rest.
And I find it is easier to read on a separate line.

File: src/compiler/crystal/tools/doc/macro.cr (hunk 4 above)
Comment (j8r) [reply src/compiler/crystal/tools/doc/macro.cr]: you can also do
```crystal
io << to_s.gsub(/<.+?>/, "").delete(' ') \
    << "-macro"
```
to continue the appends to `io` and not create a new one.

Comment (j8r) [reply src/compiler/crystal/tools/doc/macro.cr]: nvm, the performance here doesn't really matters as @straight-shoota said.

Comment (jhass) [reply src/compiler/crystal/tools/doc/macro.cr]: There's no performance benefit in that I don't see one style wise either.

Comment (Sija) [reply src/compiler/crystal/tools/doc/macro.cr]: @j8r Please, stop with those style-related comments. This PR is about perf improvements, not specific code style preferences.

Comment (j8r) [reply src/compiler/crystal/tools/doc/macro.cr]: here are the results
```crystal
require "benchmark"

Benchmark.ips do |x|
  x.report "one append to io" do
    999.times do
    String.build do |io|
      io << "abc" << "bcd"
    end
    end
  end

  x.report "one append to io with backslash" do
    999.times do
    String.build do |io|
      io << "abc" \
         << "bcd"
    end
    end
  end

  x.report "two appends to io" do
    999.times do
    String.build do |io|
      io << "abc"
      io << "bcd"
    end
    end
  end
end
```
results:
```
               one append to io   7.59k (131.82µs) (± 9.25%)  1.10× slower
one append to io with backslash   8.37k (119.44µs) (± 2.14%)       fastest
              two appends to io   8.37k (119.49µs) (± 2.43%)  1.00× slower
```
I'm wrong on which one is fast @Sija, but you're wrong saying that there are no speed differences.

File: src/compiler/crystal/compiler.cr (hunk 2 above)
Comment (r00ster91) [reply src/compiler/crystal/compiler.cr]: You are right. It's actually faster.
```Crystal
require "benchmark"

filename = "file.cr"

Benchmark.ips do |x|
  x.report "3 prints" do
    3.times do
      print " - "
      print filename
      print ": "
    end
  end

  x.report "interpolation print" do
    3.times do
      print " - #{filename}: "
    end
  end

  x.report "comma print" do
    3.times do
      print " - ", filename, ": "
    end
  end
end
```
(I used `3.times` to have clearer results)
```
           3 prints   2.47k (404.48µs) (±10.79%)  2.53× slower
interpolation print   5.75k (173.88µs) (±11.99%)  1.09× slower
        comma print   6.25k (160.11µs) (±13.01%)       fastest
```

But i'm sure there are many more interpolation prints like this. Does that means every interpolation print should be changed to a comma print instead?

File: src/compiler/crystal/tools/doc/macro.cr (hunk 4 above)
Comment (jhass) [reply src/compiler/crystal/tools/doc/macro.cr]: There's no difference, `131.82 * 0.092 = 12.12`, `131.82 - 12.12 = 119.7

Comment (straight-shoota) [reply src/compiler/crystal/tools/doc/macro.cr]: Your benchmark result is clearly wrong because the first two examples produce exactly the same AST and byte code. If one is faster than the other, that's because of external influences.

Comment (jhass) [reply src/compiler/crystal/tools/doc/macro.cr]: They're all the same even in the results, no point in even arguing about environmental issues these micro benchmarks have.

**Comment by RX14** [comment]
Can we please stop with the PRs trying to change such trivial stuff as this, it likely provides little measurable impact and isn't the most productive use of anyone's time.

**Comment by r00ster91** [comment]
I think this can be merged now. Or did anybody else noticed something that could maybe be improved?

**Comment by r00ster91** [comment]
Oh all the checks failed. But thats probably because CI isn't using 0.25.0 yet.

**Comment by straight-shoota** [comment]
Please remove the unrelated commit and create a new PR for that.

**Comment by straight-shoota** [comment]
@r00ster91 Why not simply reset the branch to 4fafe427437d3e8d57d0cc9e370a5b8acd0a3ff5?

**Comment by r00ster91** [comment]
@straight-shoota I'm not very experienced with git yet. I tried resetting to HEAD~{1} or something but it didn't really worked so I changed it back by hand. But next time I will remember to specify the commit ID. Thanks.

File: src/compiler/crystal/compiler.cr (hunk 2 above)
Comment (Sija) [reply src/compiler/crystal/compiler.cr]: Please, use `print " - ", filename, ": "` version.

**Comment by asterite** [comment]
I personally don't think such cosmetic changes are needed or have any benefit. Maybe they improve the performance (or not) of a small part of the program that's not a bottleneck. But it does distract developers and contributors to comment and argue on these small things, instead of having them focus on more important things (be it Crystal-related or not).

**Comment by r00ster91** [comment]
Okay I will stop now with these commits and wait for this to get merged.

**Comment by bcardiff** [comment]
Closing since there are conflicts. And the last comment states a stop in this changes.

I notice they were some changes in doc also, if they are wanted due to readability feel free to open a separate PR.

@r00ster91 please note that your contributions are very welcome and I hope you will keep them coming :-)

**Comment by j8r** [comment]
Too bad, this PR was not that far from a merge.
More than cosmetic changes, keep also in mind that the code also acts as a reference for newcomers.
@asterite Everyone, including Crystal contributors, is free to spend their time on what (s)he deems worth to do :)

**Comment by Sija** [comment]
I'd suggest finishing and merging this PR. Cosmetic or not, all in all they improve code readability and performance so they worthwhile enough.

**Comment by RX14** [comment]
> Cosmetic or not, all in all they improve code readability and performance so they worthwhile enough

While they do improve performance (imperceptibly) and readability (imperceptibly), these PRs use up review bandwidth as well. You're free to make them, but I don't think it's worth my time to review such PRs too, especially when they end up with 14 commits and about 10 reviews such as this one.
